ds701_book/24-NN-II-Backprop.qmd

```diff
@@ -145,79 +120,84 @@ plt.grid()
 
-> So Logistic Regression is similar to one neuron with a sigmoid activation function.
+::: {.callout-important}
+So Logistic Regression is similar to one neuron with a sigmoid activation
+function.
+:::
 
-```{python}
-#| hide_input: true
-#| slideshow: {slide_type: subslide}
-#| tags: [hide-input]
-alphas = [-4, -8,-12,-20, 0]
-betas = [0.4,0.4,0.6,1, 1]
-x = np.arange(-10,35)
-fig = plt.figure(figsize=(8, 6)) 
-ax = plt.subplot(111)
+## Multi-Layer Perceptron (MLP) or Fully Connected Network (FCN)
 
-for i in range(len(alphas)):
-    a = alphas[i]
-    b = betas[i]
-    y = np.exp(a+b*x)/(1+np.exp(a+b*x))
-    ax.plot(x,y,label=r"$\alpha=%d,$    $\beta=%3.1f$" % (a,b))
 
-ax.tick_params(labelsize=12)
-ax.set_xlabel('x', fontsize = 14)
-ax.set_ylabel('$p(x)$', fontsize = 14)
-ax.legend(loc='center left', bbox_to_anchor=(1, 0.5), prop={'size': 16})
-ax.set_title('Logistic Functions', fontsize = 16);
-```
+![](figs/NN-figs/L24-fcn-dag.png){width="75%"}
 
-### Multi-Layer Perceptron (MLP) or Fully Connected Network (FCN)
-
-<center>
-    
-<img src="figs/NN-figs/neural_net2.jpeg" width="75%">
-    
-</center>
-
-
-From [cs231n](https://cs231n.github.io/convolutional-networks/)
+From [Understanding Deep Learning](https://udlbook.github.io).
 
 Multiple artificial neurons can be acting on the same inputs, in what we call
 a _layer_, and we can have more than one _layer_ until we produce one or more
 outputs.
 
-The example above shows a network with _3 inputs_, two layers of neurons, each
-with 4 neurons, followed by one layer that produces a single value output.
+The example above shows a network with _3 inputs_, three hidden layers of 4, 2 and 3 neurons, respectively, followed by one layer that produces two continuous value outputs.
+
+E.g. a regression problem.
+
+## FCN Matrix Equations
+
+We can express the matrix equations for a $K$ layer network as
+
+$$
+\begin{aligned}
+\mathbf{h}_1 &= \mathbf{a}[\mathbf{\beta}_0 + \mathbf{\Omega}_0 \mathbf{x}] \\
+\mathbf{h}_2 &= \mathbf{a}[\mathbf{\beta}_1 + \mathbf{\Omega}_1 \mathbf{h}_1] \\
+\vdots \\
+\mathbf{h}_K &= \mathbf{a}[\mathbf{\beta}_{K-1} + \mathbf{\Omega}_{K-1} \mathbf{h}_{K-1}] \\
+\mathbf{\hat{y}} &= \mathbf{\beta}_K + \mathbf{\Omega}_K \mathbf{h}_K \\
+\end{aligned}
+$$
+
+And we can write the square loss function in vector notation as
+
+$$
+L = \sum_{i=1}^N \ell_i = \sum_{i=1}^N \|\mathbf{y}_i - \mathbf{\hat{y}}_i\|_2^2
+$$
+
+which is calculated for each training example $i$ and then accumulated.
+
+So what we want to do is find the partial derivatives of $L$ with respect to each parameter in the network.
+
+In other words, we want to compute
+
+$$
+\frac{\partial L}{\partial \mathbf{\beta}_k}, \frac{\partial L}{\partial \mathbf{\Omega}_k}, \frac{\partial L}{\partial \mathbf{\beta}_{k-1}}, \frac{\partial L}{\partial \mathbf{\Omega}_{k-1}}, \ldots, \frac{\partial L}{\partial \mathbf{\beta}_0}, \frac{\partial L}{\partial \mathbf{\Omega}_0}
```

Comment:
We haven't mentioned derivatives with respect to vectors or matrices, is it worth remarking here that there are well-defined rules that govern how these are calculated and that, e.g., pytorch, handle these calculations for you?

Reply:
Added

In the notation above, $\mathbf{\beta}_k$ and $\mathbf{\Omega}_k$ are the bias
vectors and weight matrices of the $k$-th layer. Just as with partial derivatives
with respect to scalars, there are well-defined rules for how to compute the
partial derivatives of vector and matrix expressions and, in fact, PyTorch and
TensorFlow implement these calculations.